src/expr/scalar/mod.rs

```diff
@@ -306,7 +306,7 @@ impl ScalarExpr {
                         Ok(regex) => expr1.take().call_unary(UnaryFunc::MatchRegex(Regex(regex))),
                         Err(_) => ScalarExpr::literal_null(e.typ(&relation_type)),
                     };
-                } else if *func == BinaryFunc::DateTrunc && expr1.is_literal() {
+                } else if *func == BinaryFunc::DateTruncTimestamp && expr1.is_literal() {
```

Comment:
This reduction should be applied to both `DateTruncTimestamp` and `DateTruncTimestampTz`. That's going to be a bit annoying, I'm afraid.

One option would be to do what we do for extract and have two copies of date_trunc_{milliseconds,second,minute,...}: one for timestamp and one for timestamptz.

A nicer option, I think, would be to promote the `scalar::func::format::DateTime` trait to `scalar::func::DateTime`, and then make all of the extract and date_trunc functions as generic over a `DateTime` implementation. Then you'd just need to adjust `BinaryFunc::eval` and `UnaryFunc::eval` to call `unwrap_timestamp` or `unwrap_timestamptz` as appropriate, and let `rustc` select the right generic implementation.

Reply:
Alright -- I think I've done what you asked - let me know what you think. I think this is failing some assertions on explain plan + I'll clean up commits and push out tomorrow